### ISSUE TITLE
Another problem with "Gino engine is not initialized."

### ISSUE THREAD

**Issue by ShagonRU** [fantix/gino#535]
* GINO version: 0.8.3
* Python version: 3.7.4
* asyncpg version: 0.18.3
* aiocontextvars version: 0.2.2
* PostgreSQL version: 11.4

### Description

I've tried to use GINO in aiohttp through gino.ext.aiohttp. For ~5 hours i've tried to fix problems by myself, but i failed. Every time i get "gino.exceptions.UninitializedError: Gino engine is not initialized."
So i tried to use examples from https://github.com/fantix/gino/blob/master/tests/test_aiohttp.py (so i've copied this test and models and it crashes with async_generator related errors), but they are did not work too, i used different combinations of methods from this file, docs and other issues

### What I Did
There are a clear code version without random moves which i does while trying to solve problem, it is as close as possible to examples from documentation and tests.

__main__.py (using it to launch application from docker)
```python
def main() -> None:
    app = init_app()
    app_settings = app['config']['app']
    aiohttp_swagger.setup_swagger(app)


    web.run_app(
        app,
        host=app_settings['host'],
        port=app_settings['port'],
    )

if __name__ == '__main__':
    main()
```

app.py
```python
from gino.ext.aiohttp import Gino

def init_app(config: Optional[List[str]] = None) -> web.Application:
    db = Gino()
    app = web.Application(middlewares=[error_middleware, db])

    init_config(app, config=config)
    init_routes(app)

    app['config']['gino'] = {
        'dsn': 'postgresql://{user}:{password}@{host}:{port}/{database}'.format(**app['config']['postgres'])
    }
    db.init_app(app)

    return app
```

views.py
```python
@auth_routes.get('/root/')
async def root(request: web.Request) -> web.Response:
    from user_repo import setup_root_user
    await setup_root_user(request)
```

user_repo.py
```python
async def setup_root_user(request):
    await UserIdentity.create(id=uuid.uuid4(), login='root', password_hash=pbkdf2_sha256.hash('root'))
```

models.py
```python
from gino.ext.aiohttp import Gino

db = Gino()

class UserIdentity(db.Model):
    __tablename__ = 'user_identity'

    id = db.Column(db.Unicode(), primary_key=True)
    login = db.Column(db.Unicode(), unique=True)
    password_hash = db.Column(db.Unicode())
```

And i trying to use "request['connection'].first_or_404(user_reate_query.query)", but i just get following exception before any request, i do not even have time to use request['connection'] which is stored in app.
I even tried to place "await db.gino.create_all()", bind, create engine in views.py, in user_repo, but every time get UninitializedError while tried to do a request to db. I evet get got a funny situation when I saw "<gino.ext.aiohttp.GinoEngine object at 0x0000012CE265D388>" from print and "gino.exceptions.UninitializedError: Gino engine is not initialized" from next line query.
I've set a debug point inside of a "after_server_stop" method, and I have never visited this method, so I can decide that I do not call init_app more than 1 time and connection can't be closed.

I have no idea what i should do.

Example of traceback:
```python
======== Running on http://0.0.0.0:8080 ========
(Press CTRL+C to quit)
print inside user_repo <gino.ext.aiohttp.GinoConnection object at 0x00000136D02699C8>
print inside user_repo <class 'gino.ext.aiohttp.GinoConnection'>
Error handling request
Traceback (most recent call last):
  File "C:\Users\Shagon\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\aiohttp\web_protocol.py", line 418, in start
    resp = await task
  File "C:\Users\Shagon\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\aiohttp\web_app.py", line 458, in _handle
    resp = await handler(request)
  File "C:\Users\Shagon\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\aiohttp\web_middlewares.py", line 119, in impl
    return await handler(request)
  File "C:\Users\Shagon\PycharmProjects\nwwh\newwheel_backend\middlewares.py", line 7, in error_middleware
    response = await handler(request)
  File "C:\Users\Shagon\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\gino\ext\aiohttp.py", line 113, in _middleware
    return await handler(request)
  File "C:\Users\Shagon\PycharmProjects\nwwh\newwheel_backend\auth\views.py", line 104, in root
    await setup_root_user(request)
  File "C:\Users\Shagon\PycharmProjects\nwwh\newwheel_backend\repository\user_repo.py", line 27, in setup_root_user
    await UserIdentity.create(id=uuid.uuid4(),login='root', password_hash=pbkdf2_sha256.hash('root'))
  File "C:\Users\Shagon\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\gino\crud.py", line 441, in _create_without_instance
    return await cls(**values)._create(bind=bind, timeout=timeout)
  File "C:\Users\Shagon\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\gino\crud.py", line 473, in _create
    row = await bind.first(q)
  File "C:\Users\Shagon\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\gino\api.py", line 501, in __getattribute__
    raise self._exception
gino.exceptions.UninitializedError: Gino engine is not initialized.
```

**Comment by wwwjfy** [comment]
This seems to be the problem in #530.

`db = Gino()` should be called only once. Usually we define a global `db` variable and in other files import it from app.py

**Comment by mikekeda** [comment]
for app.py and models.py you are using different db instances.
Try to move `db = Gino()` to separate file and in app.py and models.py import db from that file.

**Comment by ShagonRU** [comment]
![My projects](https://user-images.githubusercontent.com/28485306/64076606-9d3fa000-cccf-11e9-969c-5ee5eac87057.png)

Thank you!
So if i using only one instanse of `db`, i have some problems with imports. As you can see at the attached image, my project have the following structure:
```
backend
-- users
-- -- models.py
-- app.py
```

When i tried to import `db` like this:
`app.py`
```python
from gino.ext.aiohttp import Gino

__all__ = ['db', ] # tried also without this

# from newwheel_backend.db import db     # code from working way
db = Gino()

def init_app(config: Optional[List[str]] = None) -> web.Application:
    app = web.Application(middlewares=[error_middleware, db])
    init_config(app, config=config)
    init_routes(app)

    app['config']['gino'] = {
        'dsn': 'postgresql://{user}:{password}@{host}:{port}/{database}'.format(**app['config']['postgres'])
    }
    db.init_app(app)

    return app
```
`models.py`
```python
__all__ = ['User', 'UserIdentity', ]

# from newwheel_backend.db import db     # code from working way
from newwheel_backend.app import db

class UserIdentity(db.Model):
    __tablename__ = 'user_identity'

    id = db.Column(db.Unicode(), primary_key=True)
    login = db.Column(db.Unicode(), unique=True)
    password_hash = db.Column(db.Unicode())
```

App don't starts, because in `models.py` i getting 
`ImportError: cannot import name 'db' from 'newwheel_backend.app' (C:\Users\Shagon\PycharmProjects\nwwh\newwheel_backend\app.py)`
BUT if i declare variable `db = Gino()` at the different file in same level with `app.py`:
```
backend
-- users
-- -- models.py
-- app.py
-- db.py
```

The only code inside of `db.py`:
```python
from gino.ext.aiohttp import Gino
db = Gino()
```

and in `app.py`
```python
from newwheel_backend.db import db
# instead of
# db = Gino()
```
in `models.py`
```python
from newwheel_backend.db import db
# instead of
# from newwheel_backend.app import db
```

There are no error, but `app.py` and `db.py` at the same project level, i don't even now, why it works by this way =/

# == == == == ==

And another issue with creating tables
Only way that i find is to add function which contains `db.gino.create_all()` in `app.on_startup`, by this way tables are creating after binding, my signal was added after gino's `after_server_starts`, it works because
>Signal handlers will be run sequentially, in order they were added.
  
  as it was said in the aiohttp docs

Code is has to be like  
`app.py`
```python
    app['config']['gino'] = {
        'dsn': 'postgresql://{user}:{password}@{host}:{port}/{database}'.format(**app['config']['postgres'])
    }
    db.init_app(app)

    from newwheel_backend.setup_tables import setup_tables
    app.on_startup.append(setup_tables)
```

`setup_tables.py`
```python
from newwheel_backend.db import db


async def setup_tables(app) -> None:
    await db.gino.create_all()
```

it works, tables was created, i checkd them in pgAdmin.

But is there a smarter way to create tables? For example, writing a function that connects to the database will somehow create already written models (???) and close the connection; but there are problem because it will be a different instance of db, again a problem(or not, it will work?). Run this function from the docker-compose before starting the server.
I can’t think of something more reasonable that I won’t try to create tables every time the server starts. 

Thanks so much for answering the questions!

**Comment by wwwjfy** [comment]
I suspect the import issue you have is because of the order of import. Can you try to print debug strings before declarations to check?

In a pragmatic way, database initialization should be separated from the server. You could try [alembic](http://alembic.sqlalchemy.org) to manage migrations. Run it every time before starting server.

**Comment by ShagonRU** [comment]
@wwwjfy Which import i should move up and print? Now I will try to change the order of imports

So, Is there a native way to use uuid without converting to a string via str()? If I use pure uuid.uuid4() in while creating user, even if i declare column as db.String() or db.Unicode(), I get an asyncpg driver error
`asyncpg.exceptions.DataError: invalid input for query argument $1: UUID('6a6de88e-9d00-4d05-b46c-a7594f758c... (expected str, got UUID)`
but if i convert it using str(uuid.uuid4()), there are no error. Is it ok?

**Comment by wwwjfy** [comment]
To verify the idea, you can try to `print('init db')` in app.py right before `db = Gino()` and `print('init UserIdentity')` right before the class declaration.

UUID itself is not a string, you may use UUID.hex to get a string.

I suggest opening new issue for different questions, to clearer context. Or you can use https://gitter.im/python-gino/Lobby for usage questions.

**Comment by ShagonRU** [comment]
@wwwjfy it works only if i place it at very top of `app.py` or higer then module-level imports
```python
from gino.ext.aiohttp import Gino
db = Gino()
```
but i get warning from pycharm `that if I open too much issue, this will clutter up the section` (of course, i can ignore it)
If i move any string - db or import from gino.ext.aiohttp - i will get import error again.

working case:
![изображение](https://user-images.githubusercontent.com/28485306/64077900-7ab48380-ccdd-11e9-93b2-6930a2137346.png)


Sorry for so many comments, i thought that if I open too much issue, this will clutter up the issue section.
After resolving problem from this comment you can close my issue if you think it should be closed.
Anyway, thank you for your help!

**Comment by wwwjfy** [comment]
So this is actually the cause. As coming to how to organize packages, there is no right answer, either using a db.py or suppressing the pycharm warning, or maybe other ways should be your call. :)

No worries about too many issues, as others may also meet the same ones.
I'll close this issue since it's all figured out.
